src/RamIndexAnalysis.h

```diff
@@ -34,16 +34,160 @@
 #include <utility>
 #include <vector>
 
-#define RIA_NIL 0
-#define RIA_INF -1
-
 // define if enable unit tests
 #define M_UNIT_TEST
 
 namespace souffle {
 
 class RamTranslationUnit;
 
+/** search signature of a RAM operation; each bit represents an attribute of a relation.
+ * A one represents that the attribute has an assigned value; a zero represents that
+ * no value exists (i.e. attribute is unbounded) in the search. */
+class SearchSignature {
+public:
+    explicit SearchSignature(size_t arity, uint64_t mask) : bits(arity, false) {
+        size_t len = bits.size();
+        for (size_t i = 0; i < len; ++i) {
+            bits[i] = (mask % 2);
+            mask /= 2;
+        }
+    }
+    // array subscript operator
+    inline bool operator[](std::size_t pos) const {
+        assert(pos < bits.size());
+        return bits[pos];
+    }
+
+    // comparison operators
+    inline bool operator<(const SearchSignature& other) const {
+        assert(bits.size() == other.bits.size());
+        size_t len = bits.size();
+        for (size_t i = 0; i < len; ++i) {
+            size_t index = len - i - 1;  // get the most significant bit
+
+            // if our bit is not set and other's bit is set then it is smaller
+            if (bits[index] < other.bits[index]) {
+                return true;
+            }
+            // if other's bit is set and ours is not set then it is larger
+            else if (bits[index] > other.bits[index]) {
+                return false;
+            }
+        }
+        return false;
+    }
+    inline bool operator>(const SearchSignature& other) const {
+        return other < *this;
+    }
+    inline bool operator==(const SearchSignature& other) const {
+        assert(bits.size() == other.bits.size());
+        return bits == other.bits;
+    }
+    inline bool operator!=(const SearchSignature& other) const {
+        return !(other == *this);
+    }
```

Comment:
Potentially confusing behaviour: What happens in the case where `this->bits` and `rhs.bits` have different lengths? e.g. `10` vs. `100`? The details of the backing store should be opaque to the user, does it canonicalize after every operation? This is a bit-set and should behave as if it were a `set<int>` IMO.